### PR TITLE
[SYCL] Improve error handling for kernel invocation

### PR DESCRIPTION
**Pull request by MochalovaAn**
Improve error handling in enqueue_kernel_launch::handleError intel/llvm#935

### CODE REVIEW

File: sycl/include/CL/sycl/detail/pi.h [hunk 1]
```diff
@@ -53,27 +53,31 @@ typedef pi_uint64   pi_bitfield;
...
-  PI_INVALID_BINARY               = CL_INVALID_BINARY,
-  PI_INVALID_QUEUE                = CL_INVALID_COMMAND_QUEUE,
-  PI_OUT_OF_HOST_MEMORY           = CL_OUT_OF_HOST_MEMORY,
-  PI_INVALID_PROGRAM              = CL_INVALID_PROGRAM,
-  PI_INVALID_MEM_OBJECT           = CL_INVALID_MEM_OBJECT,
-  PI_OUT_OF_RESOURCES             = CL_OUT_OF_RESOURCES,
-  PI_INVALID_EVENT                = CL_INVALID_EVENT,
-  PI_INVALID_EVENT_WAIT_LIST      = CL_INVALID_EVENT_WAIT_LIST,
+  PI_SUCCESS = CL_SUCCESS,
+  PI_RESULT_INVALID_KERNEL_NAME = CL_INVALID_KERNEL_NAME,
```
Comment (mlychkov) [inline sycl/include/CL/sycl/detail/pi.h]: May we align this code spelling with other codes? I mean
```suggestion
  PI_INVALID_KERNEL_NAME = CL_INVALID_KERNEL_NAME,
```

**Comment by romanovvlad** [comment]
Could you please explain what is wrong with current error handling and how exactly you are improving it?

**Comment by MochalovaAn** [comment]
As stated in this issue (github.com/intel/llvm/issues/935) , the current handling covers only CL_INVALID_WORK_GROUP_SIZE error code . In this PR there is processing for:

- CL_INVALID_KERNEL_ARGS
- CL_INVALID_WORK_ITEM_SIZE
- CL_MISALIGNED_SUB_BUFFER_OFFSET
- CL_IMAGE_FORMAT_NOT_SUPPORTED
- CL_MEM_OBJECT_ALLOCATION_FAILURE

File: sycl/source/detail/error_handling/enqueue_kernel.cpp [hunk 2]
```diff
@@ -210,14 +210,71 @@ bool handleInvalidWorkGroupSize(const device_impl &DeviceImpl, pi_kernel Kernel,
...
+
+  const plugin &Plugin = DeviceImpl.getPlugin();
+  RT::PiDevice Device = DeviceImpl.getHandleRef();
+
+  size_t MaxWISize[] = {0, 0, 0};
+
+  Plugin.call<PiApiKind::piDeviceGetInfo>(Device, PI_DEVICE_MAX_WORK_ITEM_SIZE,
+                                          sizeof(MaxWISize), &MaxWISize,
+                                          nullptr);
+  for (int i = 0; i < NDRDesc.Dims; i++) {
```
Comment (romanovvlad) [inline sycl/source/detail/error_handling/enqueue_kernel.cpp]:
```suggestion
  for (int I = 0; i < NDRDesc.Dims; i++) {
```

Comment (MochalovaAn) [reply sycl/source/detail/error_handling/enqueue_kernel.cpp]: What is this change for?

Comment (mlychkov) [reply sycl/source/detail/error_handling/enqueue_kernel.cpp]: This change is for compliance of your code to coding style in current project. All variables should be written in upper camel case.

File: sycl/source/detail/error_handling/enqueue_kernel.cpp [hunk 3]
```diff
@@ -210,14 +210,71 @@ bool handleInvalidWorkGroupSize(const device_impl &DeviceImpl, pi_kernel Kernel,
...
+
+  size_t MaxWISize[] = {0, 0, 0};
+
+  Plugin.call<PiApiKind::piDeviceGetInfo>(Device, PI_DEVICE_MAX_WORK_ITEM_SIZE,
+                                          sizeof(MaxWISize), &MaxWISize,
+                                          nullptr);
+  for (int I = 0; I < NDRDesc.Dims; I++) {
+    if (NDRDesc.LocalSize[I] > MaxWISize[I])
+      throw sycl::nd_range_error(
+          "Number of work-items in a work-group exceed limit for dimension "
```
Comment (AlexeySachkov) [inline sycl/source/detail/error_handling/enqueue_kernel.cpp]:
```suggestion
          "Number of work-items in a work-group exceed limit of given device for dimension "
```

File: sycl/source/detail/error_handling/enqueue_kernel.cpp [hunk 4]
```diff
@@ -210,14 +210,71 @@ bool handleInvalidWorkGroupSize(const device_impl &DeviceImpl, pi_kernel Kernel,
...
+        "when the sub-buffer object is created is not aligned "
+        "to CL_DEVICE_MEM_BASE_ADDR_ALIGN value for device associated"
+        " with queue",
+        PI_MISALIGNED_SUB_BUFFER_OFFSET);
+
+  case PI_MEM_OBJECT_ALLOCATION_FAILURE:
+    throw sycl::nd_range_error(
+        "failure to allocate memory for data store associated with image"
+        " or "
+        "buffer objects specified as arguments to kernel",
```
Comment (AlexeySachkov) [inline sycl/source/detail/error_handling/enqueue_kernel.cpp]:
```suggestion
        " or buffer objects specified as arguments to kernel",
```

File: sycl/source/detail/error_handling/enqueue_kernel.cpp [hunk 5]
```diff
@@ -210,14 +210,71 @@ bool handleInvalidWorkGroupSize(const device_impl &DeviceImpl, pi_kernel Kernel,
...
     return handleInvalidWorkGroupSize(DeviceImpl, Kernel, NDRDesc);
-  // TODO: Handle other error codes
+
+  case PI_INVALID_KERNEL_ARGS:
+    throw sycl::nd_range_error(
+        "The kernel argument values have not been specified "
+        " OR "
+        "a kernel argument declared to be a pointer to a type"
+        " does not point to a named address space",
+        PI_INVALID_KERNEL_ARGS);
```
Comment (AlexeySachkov) [inline sycl/source/detail/error_handling/enqueue_kernel.cpp]: I see that I suggested to handle this error code in #935, but at the same time the second half of this message is pretty confusing for the SYCL user - he doesn't declare arguments for kernel explicitly: this is done automatically by compiler based on labda capture-list or members of function object used as kernel.

However, in SYCL, this error is also triggered sometimes if kernel captured a non-USM pointer on GPU (I saw several examples with `this` pointer), so, I suggest to rewrite the second part of this message to something like "a kernel captured non-USM pointer"

File: sycl/source/detail/error_handling/enqueue_kernel.cpp [hunk 6]
```diff
@@ -210,14 +210,71 @@ bool handleInvalidWorkGroupSize(const device_impl &DeviceImpl, pi_kernel Kernel,
...
+
+  case PI_INVALID_WORK_ITEM_SIZE:
+    return handleInvalidWorkItemSize(DeviceImpl, NDRDesc);
+
+  case PI_IMAGE_FORMAT_NOT_SUPPORTED:
```
Comment (AlexeySachkov) [inline sycl/source/detail/error_handling/enqueue_kernel.cpp]: Please handle `CL_INVALID_IMAGE_SIZE` as well

**Comment by MochalovaAn** [comment]
https://github.com/intel/llvm/pull/1339

**Comment by bader** [comment]
@MochalovaAn, please, resolve conflicts.

File: sycl/source/detail/error_handling/enqueue_kernel.cpp [hunk 7]
```diff
@@ -211,14 +211,76 @@ bool handleInvalidWorkGroupSize(const device_impl &DeviceImpl, pi_kernel Kernel,
...
+  size_t MaxWISize[] = {0, 0, 0};
+
+  Plugin.call<PiApiKind::piDeviceGetInfo>(
+      Device, PI_DEVICE_INFO_MAX_WORK_ITEM_SIZES, sizeof(MaxWISize), &MaxWISize,
+      nullptr);
+  for (int i = 0; i < NDRDesc.Dims; i++) {
+    if (NDRDesc.LocalSize[i] > MaxWISize[i])
+      throw sycl::nd_range_error(
+          "Number of work-items in a work-group exceed limit for dimension "
+          "{I}: {NDRDesc.LocalSize[I]} > {MaxWISize[I]}",
```
Comment (AlexeySachkov) [inline sycl/source/detail/error_handling/enqueue_kernel.cpp]: BTW, why do you need `{}` here?

File: sycl/source/detail/error_handling/enqueue_kernel.cpp [hunk 8]
```diff
@@ -177,7 +178,6 @@ bool handleInvalidWorkGroupSize(const device_impl &DeviceImpl, pi_kernel Kernel,
 
   const plugin &Plugin = DeviceImpl.getPlugin();
   RT::PiDevice Device = DeviceImpl.getHandleRef();
-
```
Comment (AlexeySachkov) [inline sycl/source/detail/error_handling/enqueue_kernel.cpp]: Was this change done intentionally? If no, I suggest to revert it. The same applies to a new line above (169)

**Comment by bader** [comment]
> I would recommend to squash all commits into one manually before merge.

Why?

**Comment by mlychkov** [comment]
> > I would recommend to squash all commits into one manually before merge.
> 
> Why?

Because otherwise all commits will be squashed into one automatically with merging all commit messages into one big message with useless 'Apply remarks' lines.

**Comment by bader** [comment]
> > > I would recommend to squash all commits into one manually before merge.
> > 
> > 
> > Why?
> 
> Because otherwise all commits will be squashed into one automatically with merging all commit messages into one big message with useless 'Apply remarks' lines.

I use either first commit message or PR description for the squashed commit message or combining the information from the commit messages/description, so there is no need to "squash manually" if it's just to create a commit message for the squashed commit.
Use PR description if first commit message is not good enough.

**Comment by mlychkov** [comment]
> > > > I would recommend to squash all commits into one manually before merge.
> > > 
> > > 
> > > Why?
> > 
> > 
> > Because otherwise all commits will be squashed into one automatically with merging all commit messages into one big message with useless 'Apply remarks' lines.
> 
> I use either first commit message or PR description for the squashed commit message or combining the information from the commit messages/description, so there is no need to "squash manually" if it's just to create a commit message for the squashed commit.
> Use PR description if first commit message is not good enough.

If it is so then ok, forget my comment.
I wrote it just because one of my PRs had main commit and several 'Apply remarks' commits and after merge they were squashed into one with all commit messages.

**Comment by bader** [comment]
> > > > > I would recommend to squash all commits into one manually before merge.
> > > > 
> > > > 
> > > > Why?
> > > 
> > > 
> > > Because otherwise all commits will be squashed into one automatically with merging all commit messages into one big message with useless 'Apply remarks' lines.
> > 
> > 
> > I use either first commit message or PR description for the squashed commit message or combining the information from the commit messages/description, so there is no need to "squash manually" if it's just to create a commit message for the squashed commit.
> > Use PR description if first commit message is not good enough.
> 
> If it is so then ok, forget my comment.
> I wrote it just because one of my PRs had main commit and several 'Apply remarks' commits and after merge they were squashed into one with all commit messages.

Might happen by mistake, but then to err is human.

**Pull request merged**